File: cv_pkg/src/cv_face_detect_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/msg/image.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/objdetect.hpp>

rclcpp::Node::SharedPtr node;
cv::CascadeClassifier face_cascade;
cv::Mat gray;
std::vector<cv::Rect> faces;

void Cam_RGB_Callback(const sensor_msgs::msg::Image::SharedPtr msg)
{
    cv_bridge::CvImagePtr cv_ptr;
    try
    {
        cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
    }
    catch (cv_bridge::Exception& e)
    {
        RCLCPP_ERROR(node->get_logger(), "cv_bridge exception: %s", e.what());
        return;
    }

    cv::cvtColor(cv_ptr->image, gray, cv::COLOR_BGR2GRAY);
    cv::equalizeHist(gray, gray);


    // detect face and draw rectangle
    face_cascade.detectMultiScale(
        gray,
        faces,
        1.1,          // scale factor
        3,            // min neighbors
        0,
        cv::Size(30, 30)
    );

    if (faces.size() > 0)
    {
        for (const auto& face : faces)
        {
            cv::rectangle(
                cv_ptr->image,
                face,
                cv::Scalar(0, 0, 255),
                2
            );

        }
    }

    cv::imshow("Face_Detect", cv_ptr->image);
    cv::waitKey(1);
}

int main(int argc, char const *argv[])
{
    rclcpp::init(argc, argv);

    printf("Hello World, I am the cv face detect node!\n");

    node = rclcpp::Node::make_shared("cv_face_detect_node");

    if(!face_cascade.load("/root/ros2_ws/src/cv_pkg/models/haarcascade_frontalface_default.xml"))
    {
        RCLCPP_ERROR(node->get_logger(), "Could not load face cascade model.");
        return 0;
    }

    std::shared_ptr<rclcpp::Subscription<sensor_msgs::msg::Image>> rgb_sub =
        node->create_subscription<sensor_msgs::msg::Image>(
            "/kinect2/qhd/image_raw",
            1,
            Cam_RGB_Callback
        );

    cv::namedWindow("Face_Detect");
    rclcpp::spin(node);
    cv::destroyAllWindows();
    rclcpp::shutdown();

    return 0;
}

File: cv_pkg/src/cv_face_detect_node.py
#!/usr/bin/env python3
#conding: utf-8
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2

bridge = None
node = None
face_cascade = None

def cam_rgb_callback(msg):
    global bridge, node, face_cascade

    try:
        cv_image = bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
    except Exception as e:
        node.get_logger().error(f"cv_bridge exception: {e}")
        return

    # face detection and drawing

    gray = cv2.cvtColor(cv_image, cv2.COLOR_BGR2GRAY)

    faces = face_cascade.detectMultiScale(
        gray,
        scaleFactor = 1.1,
        minNeighbors = 3,
        minSize = (30, 30)
    )

    for (x, y, w, h) in faces:
        cv2.rectangle(
            cv_image,
            (x, y),
            (x + w, y + h),
            (0, 0, 255),
            2
        )

    # Display the image using OpenCV
    cv2.imshow("Face_Detect", cv_image)
    cv2.waitKey(1)


def main(args=None):
    global bridge, node, face_cascade

    rclpy.init(args=args)

    print("Hello World, I am the cv face detect node![Python]\n")

    node = Node("cv_face_detect_node")

    face_cascade = cv2.CascadeClassifier()

    if not  face_cascade.load("/root/ros2_ws/src/cv_pkg/models/haarcascade_frontalface_default.xml"):
        node.get_logger().error("Could not load face cascade model.")
        return 0


    bridge = CvBridge()

    # create subscriber for RGB image
    node.create_subscription(
        Image,
        "/kinect2/qhd/image_raw",
        cam_rgb_callback,
        1
    )



    # creat a window
    cv2.namedWindow("Face_Detect")
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()



if __name__ == "__main__":
    main()
